object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/CalledMethodsPredicateEvaluator.java

```diff
@@ -1,23 +1,181 @@
+        theBlock.getStatements().get(0).asReturnStmt().getExpression().get();
+
+    return expressionToBooleanFormula(theExpression, booleanFormulaManager);
+  }
+
+  /**
+   * Convert between Javaparser expression and boolean formula via recursive descent.
+   *
+   * @throws UnsupportedOperationException if an unexpected expression is encountered
+   */
+  private static BooleanFormula expressionToBooleanFormula(
+      com.github.javaparser.ast.expr.Expression theExpression,
+      BooleanFormulaManager booleanFormulaManager) {
+    if (theExpression.isNameExpr()) {
+      return booleanFormulaManager.makeVariable(theExpression.asNameExpr().getNameAsString());
+    } else if (theExpression.isBinaryExpr()) {
+      if (theExpression.asBinaryExpr().getOperator().equals(BinaryExpr.Operator.OR)) {
+        return booleanFormulaManager.or(
+            expressionToBooleanFormula(
+                theExpression.asBinaryExpr().getLeft(), booleanFormulaManager),
+            expressionToBooleanFormula(
+                theExpression.asBinaryExpr().getRight(), booleanFormulaManager));
+      } else if (theExpression.asBinaryExpr().getOperator().equals(BinaryExpr.Operator.AND)) {
+        return booleanFormulaManager.and(
+            expressionToBooleanFormula(
+                theExpression.asBinaryExpr().getLeft(), booleanFormulaManager),
+            expressionToBooleanFormula(
+                theExpression.asBinaryExpr().getRight(), booleanFormulaManager));
+      }
+    } else if (theExpression.isEnclosedExpr()) {
+      return expressionToBooleanFormula(
+          theExpression.asEnclosedExpr().getInner(), booleanFormulaManager);
+    } else if (theExpression.isUnaryExpr()) {
+      if (theExpression.asUnaryExpr().getOperator().equals(UnaryExpr.Operator.LOGICAL_COMPLEMENT)) {
+        return booleanFormulaManager.not(
+            expressionToBooleanFormula(
+                theExpression.asUnaryExpr().getExpression(), booleanFormulaManager));
+      }
+    }
+    throw new UnsupportedOperationException(
+        "encountered an unexpected type of expression in an "
+            + "@CalledMethodsPredicate expression: "
+            + theExpression
+            + " was of type "
+            + theExpression.getClass());
+  }
+
+  /**
+   * Evaluate the given expression if every String in {@code trueLiterals} is replaced by "true" in
+   * the boolean formula.
+   *
+   * @param expression a boolean formula in Java format, as a String
+   * @param trueLiterals the literals in the boolean expression to treat as "true"
+   * @return whether the expression evaluates to true in the context where only the literals in
+   *     trueLiterals are true, and all other literals are false
+   */
+  public static boolean evaluate(String expression, Collection<String> trueLiterals) {
 
-    for (String cmMethod : cmMethods) {
+    for (String cmMethod : trueLiterals) {
       expression = expression.replaceAll(cmMethod, "true");
     }
 
```

Comment:
On line 182, `_` is also a legal character in a Java identifier.  However, "0-9" are not legal as the first character.
On line 182, you need to use word delimiters "\b" around the symbol being replaced.
On line 186, you also need word delimiters, but the hack might not be necessary if you use the correct regular expression in the replacement.

Please add tests for `evaluate`.  I don't see any in the repository.

Reply:
I added a few, but this isn't the point of this PR. I also fixed the regex to match the actual Java identifier rules, with the exception of unicode support, which SPEL (the expression parser we're using) doesn't have. I added a commented-out test for that.

Reply:
Also, I do not need `\b` delimiters, and they're actively harmful: we want to replace e.g. `a` in `!a` or `(a || b)`

Reply:
I agree that these fixes could be in a separate PR, since they are unrelated.  I wanted to point them out, though.

What is wrong with using `\b` delimiters to avoid matching in the middle of an identifier (which is the cause of the "tfalse" bug)?  Can you give a specific example?

Reply:
If I add `\b` delimiters, the test case `Not.java` starts to fail with three unexpected `method.invocation.invalid` errors. A representative failure is line 14:

```java
class Not {

    class Foo {
        void a() {}
        void notA(@CalledMethodsPredicate("!a") Foo this) {}
    }

    void test1(Foo f) {
        f.notA(); // ERROR HERE
    }
} 
```

Reply:
SPeL throws an exception (partial call stack):

```
:-1: other: error: SourceChecker.typeProcess: unexpected Throwable (SpelEvaluationException) while processing tests/basic/CmPredicate.java; message: EL1007E: Property or field 'a' cannot be found on null
  Compilation unit: tests/basic/CmPredicate.java
  Last visited tree at line 3 column 1:
  class CmPredicate {
  Exception: org.springframework.expression.spel.SpelEvaluationException: EL1007E: Property or field 'a' cannot be found on null; Stack trace: org.springframework.expression.spel.ast.PropertyOrFieldReference.readProperty(PropertyOrFieldReference.java:213)
  org.springframework.expression.spel.ast.PropertyOrFieldReference.getValueInternal(PropertyOrFieldReference.java:104)
  org.springframework.expression.spel.ast.PropertyOrFieldReference.getValueInternal(PropertyOrFieldReference.java:91)
  org.springframework.expression.spel.ast.SpelNodeImpl.getValue(SpelNodeImpl.java:185)
  org.springframework.expression.spel.ast.OpOr.getBooleanValue(OpOr.java:56)
  org.springframework.expression.spel.ast.OpOr.getValueInternal(OpOr.java:47)
  org.springframework.expression.spel.ast.OpOr.getValueInternal(OpOr.java:37)
  org.springframework.expression.spel.ast.SpelNodeImpl.getTypedValue(SpelNodeImpl.java:114)
  org.springframework.expression.spel.standard.SpelExpression.getValue(SpelExpression.java:173)
  org.checkerframework.checker.objectconstruction.CalledMethodsPredicateEvaluator.evaluate(CalledMethodsPredicateEvaluator.java:194)
```

Reply:
What change did you make?  It's impossible to diagnose without seeing the change you made.  You could check it into a branch.

Reply:
There are several of these, in multiple test cases (both the `BasicTest` and the `EC2Test`). I printed out some of the error-triggering code, and every instance had at least one character other than the identifier in the predicate, like `!a` or `(a`.

Reply:
Try running this:
```
public class ReplaceTest {
  public static void main(String[] args) {
    String expression = "(a && abcde && true && false)";
    System.out.println(expression);
    System.out.println(expression.replaceAll("((?!true)[_a-zA-Z][_a-zA-Z0-9]*)", "false"));
    System.out.println(expression.replaceAll("((?!true)\\b[_a-zA-Z][_a-zA-Z0-9]*\\b)", "false"));
  }
}
```

Reply:
(I also filed an issue wrt `evaluate` and unicode: https://github.com/kelloggm/object-construction-checker/issues/126)

Reply:
> Try running this:
> 
> ```
> public class ReplaceTest {
>   public static void main(String[] args) {
>     String expression = "(a && abcde && true && false)";
>     System.out.println(expression);
>     System.out.println(expression.replaceAll("((?!true)[_a-zA-Z][_a-zA-Z0-9]*)", "false"));
>     System.out.println(expression.replaceAll("((?!true)\\b[_a-zA-Z][_a-zA-Z0-9]*\\b)", "false"));
>   }
> }
> ```

Be that as it may, making the same change in `CalledMethodsPredicateEvaluator` causes the tests to fail. Because it's unrelated to the substance of this PR, please file an issue and we can address it later.

Reply:
> What change did you make? It's impossible to diagnose without seeing the change you made. You could check it into a branch.

I checked it in on the `subtyping-for-cmp-with-new-regex` branch

Reply:
The problem is that you didn't use the regex that I used above, but enclosed it in "(...)*" which let it match empty strings.  I fixed the `subtyping-for-cmp-with-new-regex` branch to use the right regex and it passes CI.

Reply:
Thanks for fixing this - I knew there must have been some kind of misunderstanding.